packages/orama/src/components/sort.ts

```diff
@@ -0,0 +1,222 @@
+import { createError } from "../errors.js"
+import { ISort, OpaqueSort, Orama, Schema, SortByParams, SortConfig, SortType, SortValue } from "../types.js"
+
+interface PropertySort<K> {
+  docs: Record<string, number>
+  orderedDocs: [string, K][]
+  type: SortType
+  n: number
+}
+
+export interface Sort extends OpaqueSort {
+  sortableProperties: string[],
+  sortablePropertiesWithTypes: Record<string, SortType>
+  sorts: Record<string, PropertySort<number | string | boolean>>
+}
+
+export type DefaultSort = ISort<Sort>
+
+function innerCreate(
+  orama: Orama<{ Sort: Sort }>,
+  schema: Schema,
+  sortableDeniedProperties: string[],
+  prefix: string,
+): Sort {
+  const sort: Sort = {
+    sortableProperties: [],
+    sortablePropertiesWithTypes: {},
+    sorts: {}
+  }
+
+  for (const [prop, type] of Object.entries(schema)) {
+    const typeActualType = typeof type
+    const path = `${prefix}${prefix ? '.' : ''}${prop}`
+
+    if (sortableDeniedProperties.includes(path)) {
+      continue
+    }
+
+    if (typeActualType === 'object' && !Array.isArray(type)) {
+      // Nested
+      const ret = innerCreate(orama, type as Schema, sortableDeniedProperties, path)
+      sort.sortableProperties.push(...ret.sortableProperties)
+      sort.sorts = {
+        ...sort.sorts,
+        ...ret.sorts
+      }
+      sort.sortablePropertiesWithTypes = {
+        ...sort.sortablePropertiesWithTypes,
+        ...ret.sortablePropertiesWithTypes
+      }
+      continue
+    }
+
+    switch (type) {
+      case 'number':
+      case 'string':
+        sort.sortableProperties.push(path)
+        sort.sortablePropertiesWithTypes[path] = type
+        sort.sorts[path] = {
+          docs: {},
+          orderedDocs: [],
+          type: type,
+          n: 0
+        }
+        break
+      case 'boolean':
+      case 'boolean[]':
+      case 'number[]':
+      case 'string[]':
+        // We want to sort the arrays
+        continue
+      default:
+        console.log('INVALID_SORT_SCHEMA_TYPE', path, typeActualType)
+        throw createError('INVALID_SORT_SCHEMA_TYPE', Array.isArray(type) ? 'array' : type as unknown as string, path)
+    }
+  }
+
+  return sort
+}
+
+async function create(
+  orama: Orama<{ Sort: Sort }>,
+  schema: Schema,
+  config?: SortConfig,
+): Promise<Sort> {
+  return innerCreate(orama, schema, (config || {}).deniedProperties || [], '')
+}
+
+async function insert(
+  sort: Sort,
+  prop: string,
+  id: string,
+  value: SortValue,
+  schemaType: SortType,
+  language: string | undefined,
+): Promise<void> {
+  const s = sort.sorts[prop] as PropertySort<SortValue>
+
+  let predicate: (value: [string, SortValue]) => boolean
+  switch(schemaType) {
+    case "string":
+      predicate = (d: [string, SortValue]) => (d[1] as string).localeCompare(value as string, language) > 0
```

Comment:
For performance reason, extract these two functions and use them below using `.bind`.